LibreNMS/OS/Teldat.php

```diff
@@ -0,0 +1,139 @@
+ *
+ * @link       https://www.librenms.org
+ *
+ * @copyright  2021 Antonio Pedro Santos
+ * @author     Antonio Pedro Santos <cupidazul@gmail.com>
+ */
+
+namespace LibreNMS\OS;
+
+use App\Models\Port;
+use Illuminate\Support\Str;
+use LibreNMS\Device\WirelessSensor;
+use LibreNMS\Interfaces\Discovery\Sensors\WirelessCellDiscovery;
+use LibreNMS\Interfaces\Discovery\Sensors\WirelessClientsDiscovery;
+use LibreNMS\Interfaces\Discovery\Sensors\WirelessRsrpDiscovery;
+use LibreNMS\Interfaces\Discovery\Sensors\WirelessRsrqDiscovery;
+use LibreNMS\Interfaces\Discovery\Sensors\WirelessRssiDiscovery;
+use LibreNMS\Interfaces\Discovery\Sensors\WirelessSinrDiscovery;
+use LibreNMS\OS;
+use LibreNMS\OS\Traits\TeldatCellular;
+
+class Teldat extends OS implements
+    WirelessCellDiscovery,
+    WirelessClientsDiscovery,
+    WirelessRssiDiscovery,
+    WirelessRsrqDiscovery,
+    WirelessRsrpDiscovery,
+    WirelessSinrDiscovery
+{
+    use TeldatCellular;
+
+    private $deviceInfo = [];
+    private $hasWLANif = false;
+    private $hasGSMif = false;
+
+    public function getdevicePortsInfo()
+    {
+        $this->deviceInfo = Port::whereIn('device_id', [$this->getDeviceId()])
+                ->select('ports.device_id', 'port_id', 'ifAlias', 'ifName', 'ifDescr', 'ifOperStatus')
+                ->with(['device' => function ($query) {
+                    $query->select('device_id', 'hostname', 'sysName');
+                }])->get();
+
+        // Check if any wlan interface is up
+        foreach ($this->deviceInfo as $entIndex => $ent) {
+            if (Str::startsWith($ent['ifName'], 'wlan') and $ent['ifOperStatus'] == 'up') {
+                $this->hasWLANif = true;
+                break;
+            }
+        }
+
+        // Check if any cellular interface is up
+        foreach ($this->deviceInfo as $entIndex => $ent) {
+            if (Str::startsWith($ent['ifName'], 'cellular') and $ent['ifOperStatus'] == 'up') {
+                $this->hasGSMif = true;
+                break;
+            }
+        }
+    }
+
+    /**
+     * @return array Sensors
+     */
+    public function discoverWirelessClients()
+    {
+        $sensors = [];
+
+        if (empty($this->deviceInfo)) {
+            $this->getdevicePortsInfo();
+        }
+        if (! $this->hasWLANif) {
+            return $sensors;
+        }
+
+        $device = $this->getDeviceArray();
+        // telProdNpMonInterfWlanBSSCurrent :: Count of current associated stations
+        $data = snmpwalk_cache_oid($device, 'telProdNpMonInterfWlanBSSCurrent', [], 'TELDAT-MON-INTERF-WLAN-MIB');
+
+        // Copy deviceInfo Obj
+        $ifName = $this->deviceInfo;
+
+        // fixup incorrect/missing IfIndex mapping
+        foreach ($data as $index => $_unused) {
+            foreach ($ifName as $entIndex => $ent) {
+                $descr = $ent['ifName'];
+                unset($ifName[$entIndex]); // only use each one once
```

Comment:
tip: don't unset from the variable you are looping on inside the loop.

Reply:
Look, I just used the original foreach on the Ios.php, since the objective was to use only once, i saw no harm on keeping the original foreach as they were.

Reply:
Yeah, copying existing code is not always the path to the best code as LibreNMS has huge variance in the code within.

Anyway if you call `$ifNames = $this->cacheByIndex('IF-MIB::ifName')`

You can just do `$ifNames[$ifIndex]` to get the name.

Reply:
Ok, It would be great if I could get if ifOperStatus also, in one call like cacheByIndex( [ 'IF-MIB::ifName', IF-MIB::ifOperStatus ] )... since I'm using this info to decide if we have Wireless / Cellular interfaces up to monitor...
I'll try that and some more things...

Cheers mate. Thx.

Reply:
Looks much better now, I had some bugs also that where forcing some snmp-gets on each OID even after the snmpbulk for each sensor.

Removed APN name information from the graph name , because TELDAT devices have the APN on the "connected" peer, and if the peer changes, the name would change also, and we could loose historical data.

(re)Added PortsTable get, but included the getCacheByIndex in case it comes up empty. This solves the snmpsim tests problem.

Let's see how it pans out on the linter now...
Cheers, and Thx again...